### PR TITLE
[HUDI-1896] Intial source for Cloud Dfs

### PR DESCRIPTION
**Pull request by data-storyteller**
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.util.*;
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: nit: can we import specific class/interfaces only?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    protected static volatile Logger log = LogManager.getLogger(CloudObjectsDfsSelector.class);
+    public static HashMap<String, Boolean> map = new HashMap<>();
+
+    /**
+     * Configs supported.
+     */
+    public static class Config {
+
+        public static final String SQS_QUEUE_URL_PROP = "hoodie.deltastreamer.source.sqs.queueurl";
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: Can we rename the configs keys to be generic enough for multi clouds? SQS is AWS service, while GCP has its own. Probably, we could have an abstract class which will be extended by AWS/GCP specific classes. What do you think?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class CloudObjectsDfsSelector implements Serializable {
+
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java:92]: a serial version id would be good to have like other classes.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    protected static volatile Logger log = LogManager.getLogger(CloudObjectsDfsSelector.class);
+    public static HashMap<String, Boolean> map = new HashMap<>();
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: nit: perhaps a better name for the map to signify what it holds?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        public static final String SQS_QUEUE_MAXMESSAGESEACHBATCH_PROP = "hoodie.deltastreamer.source.sqs.max_messages_each_batch";
+        public static final String SQS_QUEUE_VISIBILITYTIMEOUT_PROP = "hoodie.deltastreamer.source.sqs.visibility_timeout_seconds";
+        public static final String SOURCE_INPUT_SELECTOR = "hoodie.deltastreamer.source.input.selector";
+    }
+
+    protected static final List<String> IGNORE_FILEPREFIX_LIST = Arrays.asList(".", "_");
+    protected static final List<String> ALLOWED_S3_EVENT_PREFIX = Arrays.asList("ObjectCreated");
+
+    //    protected final transient FileSystem fs;
+    protected final TypedProperties props;
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: Why are all these variables protected?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    protected final String queueUrl;
+    protected final AmazonSQS sqs;
+    protected final int longPollWait;
+    protected final int MaxMessagesEachRequest;
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: nit: rename to `maxMessagesEachRequest ` and `maxMessageEachBatch ` and `visibilityTimeout `

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    /**
+     * Get the messages from queue from last checkpoint.
+     *
+     * @param sourceLimit max bytes to read each time
+     * @return the list of files concatenated and their latest modified time
+     */
+    public Pair<Option<String>, String> getNextFilePathsFromQueue(Option<String> lastCheckpointStr, long sourceLimit) {
+        System.out.println("Reading messages....");
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: let's use logger

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+                JSONObject s3Object = record.getJSONObject("s3").getJSONObject("object");
+                String bucket = URLDecoder.decode(record.getJSONObject("s3").getJSONObject("bucket").getString("name"), "UTF-8");
+                String key = URLDecoder.decode(s3Object.getString("key"), "UTF-8");
+                String filePath = "s3://" + bucket + "/" + key;
+
+                String fileName = StringUtils.substringAfterLast(key, "/");
+//                System.out.println(eventTime);
+
+                // toDO - move this check to listFilesAfterCheckpoint function
+                if (IGNORE_FILEPREFIX_LIST.stream().noneMatch(fileName::startsWith)) {
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: yeah let's move this to `listFilesAfterCheckpoint ` method as you have mentioned.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+                    long fileSize = s3Object.getLong("size");
+//                    System.out.println(filePath);
+                    // we will fetch all files from queue until
+                    // we reach the max byte limit for batch
+                    if ((currentBytes + fileSize) >= sourceLimit && eventTime > newCheckpointTime) {
+                        break;
+                    }
+//                    map.put(filePath, Boolean.TRUE);
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: i'm assuming in the next revision of the PR, we will be using this map.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+            // Not fetching up-to latest message can lead to missing files
+            if (messages.isEmpty()
+//                    || (messages.size() < this.MaxMessagesEachRequest && this.longPollWait > 0)
+                    || (fetchedMessages >= this.MaxMessageEachBatch)) {
+                break;
+            }
+        }
+        return result;
+    }
+}
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: nit: line break at EOF

**Review by codope — COMMENTED** [review]
@satishmittal1111 Overall looks good. Few minor comments. Along with them, can you update the PR with the changes that we discussed?

File: hudi-utilities/pom.xml [hunk 11]
```diff
@@ -15,448 +15,455 @@
   See the License for the specific language governing permissions and
   limitations under the License.
 -->
```
Comment (codope) [inline hudi-utilities/pom.xml]: Can we undo this change? Are we adding any new dependency?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsDfsSource.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    public CloudObjectsDfsSource(TypedProperties props, JavaSparkContext sparkContext, SparkSession sparkSession,
+                                 SchemaProvider schemaProvider) {
+        super(props, sparkContext, sparkSession, schemaProvider);
+        this.pathSelector = CloudObjectsDfsSelector.createSourceSelector(props, this.sparkContext.hadoopConfiguration());
+    }
+
+    @Override
+    public Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkptStr, long sourceLimit) {
+
+        Pair<Option<String>, String> selectPathsWithLatestSQSMessage =
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsDfsSource.java]: We can rename the variable/methods to be cloud provider agnostic, e.g. rename selectPathsWithLatestSQSMessage to selectPathsWithLatestMessage

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsDfsSource.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                pathSelector.getNextFilePathsFromQueue(sparkContext, lastCkptStr, sourceLimit);
+        return selectPathsWithLatestSQSMessage.getLeft()
+                .map(pathStr -> Pair.of(Option.of(fromFiles(pathStr)), selectPathsWithLatestSQSMessage.getRight()))
+                .orElseGet(() -> Pair.of(Option.empty(), selectPathsWithLatestSQSMessage.getRight()));
+    }
+
+    private Dataset<Row> fromFiles(String pathStr) {
+        return sparkSession.read().parquet(pathStr.split(","));
+    }
+}
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsDfsSource.java]: nit: line break after the class ends.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsMetaSource.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+            return Pair.of(Option.empty(), selectPathsWithLatestSQSMessage.getRight());
+        } else {
+            return Pair.of(Option.of(fromFiles(selectPathsWithLatestSQSMessage.getLeft())), selectPathsWithLatestSQSMessage.getRight());
+        }
+    }
+
+    private Dataset<Row> fromFiles(List<String> jsonData) {
+        return sparkSession.read().json(sparkSession.createDataset(jsonData, Encoders.STRING()));
+    }
+    public static void main(String[] args) {
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsMetaSource.java]: Let's remove the main method.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+import java.util.Date;
+import java.util.Comparator;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Collections;
+import java.util.stream.Collectors;
+
+public class CloudObjectsDfsSelector implements Serializable {
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsDfsSelector.java]: I see some commonalities between this and the meta selector. Can we abstract out to a class that can be extended by both?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsMetaSelector.java [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+    private static final long serialVersionUID = 2L;
+
+    protected static volatile Logger log = LogManager.getLogger(CloudObjectsMetaSelector.class);
+
+    /**
+     * Configs supported.
+     */
+    public static class Config {
+
+        public static final String QUEUE_URL_PROP = "hoodie.deltastreamer.source.queue.url";
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsMetaSelector.java]: Few minor comments related to naming of the config keys. Let's discuss offline.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsMetaSelector.java [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+                        }
+                    }
+                } else {
+                    log.info("Message is not expected format or it's s3:TestEvent");
+                }
+                if (isMessageDelete) {
+                    ineligibleMessages.add(message);
+                }
+            }
+            if (!ineligibleMessages.isEmpty()) {
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsMetaSelector.java]: Let me know if my understanding is incorrect.
Here, we are trying to only delete ineligible messages (the ones that are not with ALLOWED_S3_EVENT_PREFIX). The eligible messages will get deleted later after the DeltaSync#writeToSink has completed. Is that correct?

**Review by codope — COMMENTED** [review]
First round looks good. Let's discuss the deletion part in a bit more detail offline.

File: hudi-utilities/pom.xml [hunk 18]
```diff
@@ -458,5 +465,27 @@
       <artifactId>junit-platform-commons</artifactId>
       <scope>test</scope>
     </dependency>
+      <dependency>
```
Comment (codope) [inline hudi-utilities/pom.xml]: Let's talk about these dependencies. It's better if we can do without them.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+  public CloudObjectsHoodieIncrSource(
+      TypedProperties props,
+      JavaSparkContext sparkContext,
+      SparkSession sparkSession,
+      SchemaProvider schemaProvider) {
+    super(props, sparkContext, sparkSession, schemaProvider);
+  }
+
+  @SuppressWarnings("checkstyle:WhitespaceAfter")
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java]: Do we need this SuppressWarnings?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+  @Override
+  public Pair<Option<Dataset<Row>>, String> fetchNextBatch(
+      Option<String> lastCkptStr, long sourceLimit) {
+
+    DataSourceUtils.checkRequiredProperties(
+        props, Collections.singletonList(Config.HOODIE_SRC_BASE_PATH));
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java:60]: Are we gonna need other properties to make the connection?

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    /** {@value #HOODIE_SRC_BASE_PATH} is the base-path for the source Hoodie table. */
+    private static final String HOODIE_SRC_BASE_PATH =
+        "hoodie.deltastreamer.source.hoodieincr.path";
+
+    /**
+     * {@value #NUM_INSTANTS_PER_FETCH} allows the max number of instants whose changes can be
+     * incrementally fetched.
+     */
+    private static final String NUM_INSTANTS_PER_FETCH =
+        "hoodie.deltastreamer.source.hoodieincr.num_instants";
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java]: Let's rename this to `hoodie.deltastreamer.source.hoodieincr.num.instants` or `hoodie.deltastreamer.source.hoodieincr.instants`. Idea is to not have underscore, if we need underscore we can replace it by `.` as well.

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+    List<String> cloudFiles = new ArrayList<>();
+    for (Row row : cloudMetaDf) {
+      String bucket = row.getString(0);
+      String key = row.getString(1);
+      String filePath = "s3://" + bucket + "/" + key;
+      cloudFiles.add(filePath);
+    }
+    String pathStr = String.join(",", cloudFiles);
+
+    // log.info("Final Schema from Source is :" + src.schema());
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/CloudObjectsHoodieIncrSource.java]: nit: remove this log

File: hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsSelector.java [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+      int maxMessagesEachRequest) {
+    List<Message> messagesToProcess = new ArrayList<>();
+
+    // Get count for available messages
+    Map<String, String> queueAttributesResult = getSqsQueueAttributes(sqs, queueUrl);
+    long approxMessagesAvailable =
+        Long.parseLong(queueAttributesResult.get("ApproximateNumberOfMessages"));
+    log.info("Approx. " + approxMessagesAvailable + " messages available in queue.");
+
+    for (int i = 0;
```
Comment (codope) [inline hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsSelector.java:144]: Let's add a comment why this loop is needed and why we can't use any collection api for partition.

**Review by codope — CHANGES_REQUESTED** [review]
@satishmittal1111 Can you resolve review comments after you have addressed them? Few more minor changes required as commented below. Please also addd test cases. 

I have one question: why are we not using SNS? My understanding is that we first publish to SNS which pushes to SQS right?

**Comment by data-storyteller** [comment]
@codope I have added the test cases. please review.

**Review by codope — COMMENTED** [review]
@satishmittal1111 Thanks for working in this. Overall, looks good to me. I'll do some minor changes and then push to a branch other than master so that it can be upstreamed.

**Pull request merged**